### PR TITLE
nxagent: half number of NameForAtom calls

### PR DESCRIPTION
**Pull request by uli42**
Because of validateString being a macro NameForAtom was called twice most of the time.

Fix that by making validateString a function.

Fixes ArticaProject/nx-libs#357

### CODE REVIEW

File: nx-X11/programs/Xserver/hw/nxagent/Atoms.c [hunk 1]
```diff
@@ -95,6 +95,10 @@ static char *nxagentAtomNames[NXAGENT_NUMBER_OF_ATOMS + 1] =
 
 static XErrorHandler previousErrorHandler = NULL;
 
+char * validateString(char *str) {
+  return str ? str : "null";
```
Comment (Ionic) [inline nx-X11/programs/Xserver/hw/nxagent/Atoms.c]: The original NX code used "(null)".

**Review by Ionic — COMMENTED** [review]
If you'd have made the function `static inline` and defined and declared it inside of `Utils.h`, we wouldn't need to pull in `Atom.h` everywhere, but in contrast just make sure that `Utils.h` is included at some point by everything using `validateString()`.

There's another problem with this approach, though, and we probably need to discuss it.

We'll generate a lot of warnings this way:

```
ionic@nopileos~% cat /tmp/chartest.c
```
```C
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#ifdef validateString_macro
#undef validateString_macro
#endif

#define validateString_macro(str) ((str) ? (str) : "(null)")

static char* validateString (char *str) {
  if (str) {
    return (str);
  }
  else {
    return ("(null)");
  }
}

static char* print_and_return (char *str) {
  if (str) {
    printf ("%s: %s\n", __PRETTY_FUNCTION__, str);
  }

  return (str);
}

static const char* print_and_return_const (const char *str) {
  if (str) {
    printf ("%s: %s\n", __PRETTY_FUNCTION__, str);
  }

  return (str);
}

int main (int argc, char **argv) {
  const char *literal = "foo";
  const char* const const_ptr_to_const_data = NULL;
  char *mutable = strdup ("bar");

  /* Silence benign warnings. */
  (void) argc;
  (void) argv;

#ifndef MACRO_CODE
  (void) validateString_macro (NULL);

  printf ("%s; %s; %s\n", validateString (print_and_return (mutable)) /* Shall work fine. */,
                          validateString (print_and_return_const (literal)) /* Not so fine... */,
                          validateString (print_and_return_const (const_ptr_to_const_data)) /* Uh oh! Now we're into dragon land for real! */);

  {
    char *not_so_const_copy = validateString (literal);
    (void) not_so_const_copy;
  }
  {
    char *not_so_const_copy = validateString (const_ptr_to_const_data);
    (void) not_so_const_copy;
  }
#else
  (void) validateString (NULL);

  printf ("%s; %s; %s\n", validateString_macro (print_and_return (mutable)) /* Shall work fine. */,
                          validateString_macro (print_and_return_const (literal)) /* Shall work fine, too. */,
                          validateString_macro (print_and_return_const (const_ptr_to_const_data)) /* And even that shall work fine! */);

  {
    char *not_so_const_copy = validateString_macro (literal);
    (void) not_so_const_copy;
  }
  {
    char *not_so_const_copy = validateString_macro (const_ptr_to_const_data);
    (void) not_so_const_copy;
  }
#endif

  free (mutable);
}
```
```
[1 running job(s)] {history#10928}  5:16:18 17-02-26
ionic@nopileos~% clang -std=c99 -Wall -Wextra -Weverything -fomit-frame-pointer -O3 -pipe -g3 -ggdb3 -gdwarf-4 /tmp/chartest.c -o /tmp/chartest -DMACRO_CODE
/tmp/chartest.c:68:11: warning: initializing 'char *' with an expression of type 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString_macro (literal);
          ^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:72:11: warning: initializing 'char *' with an expression of type 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString_macro (const_ptr_to_const_data);
          ^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

The warnings here are actually good, as the code is *really* doing something dodgy.

```
[1 running job(s)] {history#10929}  5:16:30 17-02-26
ionic@nopileos~% /tmp/chartest
char *print_and_return(char *): bar
char *print_and_return(char *): bar
const char *print_and_return_const(const char *): foo
const char *print_and_return_const(const char *): foo
bar; foo; (null)
[1 running job(s)] {history#10930}  5:16:35 17-02-26
ionic@nopileos~% clang -std=c99 -Wall -Wextra -Weverything -fomit-frame-pointer -O3 -pipe -g3 -ggdb3 -gdwarf-4 /tmp/chartest.c -o /tmp/chartest -UMACRO_CODE
/tmp/chartest.c:49:43: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
                          validateString (print_and_return_const (literal)) /* Not so fine... */,
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:50:43: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
                          validateString (print_and_return_const (const_ptr_to_const_data)) /* Uh oh! Now we're into dragon land for real! */);
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:53:47: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString (literal);
                                              ^~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:57:47: warning: passing 'const char *const' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString (const_ptr_to_const_data);
                                              ^~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
4 warnings generated.
```

"Good" warnings mixed with noise.

```
[1 running job(s)] {history#10931}  5:16:41 17-02-26
ionic@nopileos~% /tmp/chartest
char *print_and_return(char *): bar
const char *print_and_return_const(const char *): foo
bar; foo; (null)
```

There isn't an easy way out of this, as C does not support parameter overloading, though.

**Comment by uli42** [comment]
Ok, I'll fix the "null".

What about abandoning validateString for the majority of calls by integrating it into NameForAtom and using the old macro for the other places?

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mihai Moldovan" <notifications@github.com>
Gesendet: ‎26.‎02.‎2017 05:26
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Author" <author@noreply.github.com>
Betreff: Re: [ArcticaProject/nx-libs] nxagent: half number of NameForAtomcalls (#359)

@Ionic commented on this pull request.
If you'd have made the function static inline and defined and declared it inside of Utils.h, we wouldn't need to pull in Atom.h everywhere, but in contrast just make sure that Utils.h is included at some point by everything using validateString().
There's another problem with this approach, though, and we probably need to discuss it.
We'll generate a lot of warnings this way:
ionic@nopileos~% cat /tmp/chartest.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#ifdef validateString_macro
#undef validateString_macro
#endif

#define validateString_macro(str) ((str) ? (str) : "(null)")

static char* validateString (char *str) {
  if (str) {
    return (str);
  }
  else {
    return ("(null)");
  }
}

static char* print_and_return (char *str) {
  if (str) {
    printf ("%s: %s\n", __PRETTY_FUNCTION__, str);
  }

  return (str);
}

static const char* print_and_return_const (const char *str) {
  if (str) {
    printf ("%s: %s\n", __PRETTY_FUNCTION__, str);
  }

  return (str);
}

int main (int argc, char **argv) {
  const char *literal = "foo";
  const char* const const_ptr_to_const_data = NULL;
  char *mutable = strdup ("bar");

  /* Silence benign warnings. */
  (void) argc;
  (void) argv;

#ifndef MACRO_CODE
  (void) validateString_macro (NULL);

  printf ("%s; %s; %s\n", validateString (print_and_return (mutable)) /* Shall work fine. */,
                          validateString (print_and_return_const (literal)) /* Not so fine... */,
                          validateString (print_and_return_const (const_ptr_to_const_data)) /* Uh oh! Now we're into dragon land for real! */);

  {
    char *not_so_const_copy = validateString (literal);
    (void) not_so_const_copy;
  }
  {
    char *not_so_const_copy = validateString (const_ptr_to_const_data);
    (void) not_so_const_copy;
  }
#else
  (void) validateString (NULL);

  printf ("%s; %s; %s\n", validateString_macro (print_and_return (mutable)) /* Shall work fine. */,
                          validateString_macro (print_and_return_const (literal)) /* Shall work fine, too. */,
                          validateString_macro (print_and_return_const (const_ptr_to_const_data)) /* And even that shall work fine! */);

  {
    char *not_so_const_copy = validateString_macro (literal);
    (void) not_so_const_copy;
  }
  {
    char *not_so_const_copy = validateString_macro (const_ptr_to_const_data);
    (void) not_so_const_copy;
  }
#endif

  free (mutable);
}[1 running job(s)] {history#10928}  5:16:18 17-02-26
ionic@nopileos~% clang -std=c99 -Wall -Wextra -Weverything -fomit-frame-pointer -O3 -pipe -g3 -ggdb3 -gdwarf-4 /tmp/chartest.c -o /tmp/chartest -DMACRO_CODE
/tmp/chartest.c:68:11: warning: initializing 'char *' with an expression of type 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString_macro (literal);
          ^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:72:11: warning: initializing 'char *' with an expression of type 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString_macro (const_ptr_to_const_data);
          ^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
The warnings here are actually good, as the code is really doing something dodgy.
[1 running job(s)] {history#10929}  5:16:30 17-02-26
ionic@nopileos~% /tmp/chartest
char *print_and_return(char *): bar
char *print_and_return(char *): bar
const char *print_and_return_const(const char *): foo
const char *print_and_return_const(const char *): foo
bar; foo; (null)
[1 running job(s)] {history#10930}  5:16:35 17-02-26
ionic@nopileos~% clang -std=c99 -Wall -Wextra -Weverything -fomit-frame-pointer -O3 -pipe -g3 -ggdb3 -gdwarf-4 /tmp/chartest.c -o /tmp/chartest -UMACRO_CODE
/tmp/chartest.c:49:43: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
                          validateString (print_and_return_const (literal)) /* Not so fine... */,
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:50:43: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
                          validateString (print_and_return_const (const_ptr_to_const_data)) /* Uh oh! Now we're into dragon land for real! */);
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:53:47: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString (literal);
                                              ^~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
/tmp/chartest.c:57:47: warning: passing 'const char *const' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    char *not_so_const_copy = validateString (const_ptr_to_const_data);
                                              ^~~~~~~~~~~~~~~~~~~~~~~
/tmp/chartest.c:11:36: note: passing argument to parameter 'str' here
static char* validateString (char *str) {
                                   ^
4 warnings generated.
"Good" warnings mixed with noise.
[1 running job(s)] {history#10931}  5:16:41 17-02-26
ionic@nopileos~% /tmp/chartest
char *print_and_return(char *): bar
const char *print_and_return_const(const char *): foo
bar; foo; (null)
There isn't an easy way out of this, as C does not support parameter overloading, though.



In nx-X11/programs/Xserver/hw/nxagent/Atoms.c:
> @@ -95,6 +95,10 @@ static char *nxagentAtomNames[NXAGENT_NUMBER_OF_ATOMS + 1] =
 
 static XErrorHandler previousErrorHandler = NULL;
 
+char * validateString(char *str) {
+  return str ? str : "null";
The original NX code used "(null)".

**Comment by Ionic** [comment]
Or provide two functions, like I did for the printing test function. Would probably be better. We'd have to make sure that people actually use it if passing `const char*`. It looks like `MakeAtom()` actually takes `const char*`, too.

**Comment by sunweaver** [comment]
On related X.org commit:

```
commit 5623c27700b7b23a8dbbd8c8f45e5d4fa0c667e3
Author: Alan Coopersmith <alan.coopersmith@sun.com>
Date:   Mon Feb 2 19:25:14 2009 -0800

    Constify atom name strings
    
    Changes MakeAtom to take a const char * and NameForAtom to return them,
    since many callers pass pointers to constant strings stored in read-only
    ELF sections.   Updates in-tree callers as necessary to clear const
    mismatch warnings introduced by this change.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@sun.com>
    Acked-by: Peter Hutterer <peter.hutterer@who-t.net>


```

**Comment by sunweaver** [comment]
I think, we should include 5623c27700b7b23a8dbbd8c8f45e5d4fa0c667e3 in this PR. That is, weave in that X.org commit first, then fix the validateString macro/functions/whatever.

**Comment by Ionic** [comment]
You already made a partial backport in fb832b01e9a497d6e59d026cdd58068101f4c4cb, that's why `MakeAtom()` also takes `const char *` now.

**Comment by sunweaver** [comment]
ohhh bullocks, so I basically missed the rest. In the past I maybe did something wrong when investigating X.org commits...

**Comment by sunweaver** [comment]
@uli42: I can do the backporting of 5623c27700b7b23a8dbbd8c8f45e5d4fa0c667e3 tomorrow and you rebase this branch against my update branch. Ok?

**Comment by Ionic** [comment]
Yeah, I don't think you actually tried to backport the offending commit, only made `MakeAtom()` and `CopyISOLatin1Lowered()` match with upstream for some reason. Not a problem.

**Comment by uli42** [comment]
Ok

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mike Gabriel" <notifications@github.com>
Gesendet: ‎26.‎02.‎2017 15:35
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Mention" <mention@noreply.github.com>
Betreff: Re: [ArcticaProject/nx-libs] nxagent: half number of NameForAtomcalls (#359)

@uli42: I can do the backporting of 5623c27700b7b23a8dbbd8c8f45e5d4fa0c667e3 tomorrow and you rebase this branch against my update branch. Ok?

**Comment by uli42** [comment]
"null" has been replaced by "(null)"

**Comment by uli42** [comment]
I have reworked this and added to further backports to make some const warnings disappear. Please review again.

**Comment by uli42** [comment]
I have reworked this and added to further backports to make some const warnings disappear. Please review again. There was no need for splitting validateString into two functions (const/non-const).

**Comment by sunweaver** [comment]
@uli42: ping

File: nx-X11/programs/Xserver/hw/nxagent/Atoms.c [hunk 2]
```diff
@@ -95,6 +95,10 @@ static char *nxagentAtomNames[NXAGENT_NUMBER_OF_ATOMS + 1] =
 
 static XErrorHandler previousErrorHandler = NULL;
 
+const char * validateString(const char *str) {
+  return str ? str : "(null)";
+}
+
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/hw/nxagent/Atoms.c]: As earlier discussed, I second @Ionic's suggestion to put this into Utils.h as a static inline function. The validateString macro is also used for non-Atom strings.

**Review by sunweaver — APPROVED** [review]

**Comment by Ionic** [comment]
Uh... that's a very bad idea... `_XkbDupString()` is *not* equivalent to `strdup()`. Calling `strdup()` with a `NULL`-pointer as its argument is not allowed (and will most of the time lead to a SIGSEGV) - but perfectly fine with `_XkbDupString()`!

**Comment by Ionic** [comment]
Ah, but `Xstrdup()` takes care of that, I see. That's okay then.

**Pull request merged**